src/docs/binding-generics.md

```diff
@@ -437,3 +437,191 @@ We could also consider solving this on the runtime level.
 ### Generic classes
 
 TBD
+
+## Protocol constraints
+
+This section describes how to project protocol constraints on generic type parameters from Swift to C\#. The goal here is not to describe full projections of protocols; for that, please refer to the [binding protocols doc](binding-protcols.md).
+
+A Swift protocol can be mapped to C\# as an interface. Then Swift generic parameters with protocol constraint can be mapped into C\# as generic parameters with interface constraint. This should work for both functions and types.
+
+### Protocol constraints on functions
+
+Consider the following simple example in Swift:
+
+```swift
+public protocol Printable {
+ func printMe()
+}
+
+public func printPrintable<T: Printable>(data: T) {
+ data.printMe()
+}
+
+// Equivalent to: public func printPrintable<T>(data: T) where T : Printable
+```
+
+When a function is declared with a generic argument, an extra implicit argument is added to the end representing the type metadata of the generic argument (see above). If that generic argument is constrained by one or more protocols, the type metadata is followed by a pointer to the protocol witness table for each protocol. In the case of multiple protocols, e.g. T: P3 & P1 & P2, the protocol witness tables are ordered lexically by fully qualified name of the protocol. For detailed description of Metadata accessor refer to [runtime metadata doc](runtime-metadata.md).
+
+This can be projected into C\# as follows:
+
+```csharp
+public interface IPrintableProtocol
+{
+    public void PrintMe();
+    public abstract static NativeHandle GetIPrintableProtocolWitnessTable();
+
+}
+
+public static void PrintPrintable<T>(T data) where T : ISwiftObject, IPrintableProtocol
+{
+    var metadata = TypeMetadata.GetTypeMetadataOrThrow<T>();
+    var protocolWitnessTable = T.GetIPrintableProtocolWitnessTable();
+    PInvokePrintPrintable(data.GetPayload(), metadata, protocolWitnessTable);
+}
+
+[DllImport(Path, EntryPoint = "...")]
+[UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+private static extern void PInvokePrintPrintable(NativeHandle data, TypeMetadata metadata, NativeHandle protocolWitnessTable);
+```
+
+Then a type implementing the interface could look like:
+
+```csharp
+[StructLayout(LayoutKind.Sequential, Size = /* Extract size */)]
+struct FrozenStruct : ISwiftObject, IPrintableProtocol
+{
+    // Other members
+
+    public static NativeHandle GetIPrintableProtocolWitnessTable()
+    {
+        // This method extracts the witness table symbol from the dynamic library and should be cached.
+        return GetWitnessValueTable("...");
```

Comment:
To retrieve protocol witness table, protocol conformance descriptor is needed. Look at https://github.com/swiftlang/swift/blob/6221b29c6835442706fbb44b67b755d370a87d96/include/swift/Runtime/Metadata.h#L360-L370.

Reply:
Let's check if it is always available. It should be cached:
```llvm-ir
; Function Attrs: noinline nounwind memory(none)
define linkonce_odr hidden ptr @"$s7SwiftUI13_TaskModifierVAcA04ViewD0AAWl"() #8 {
entry:
  %0 = load ptr, ptr @"$s7SwiftUI13_TaskModifierVAcA04ViewD0AAWL", align 8
  %1 = icmp eq ptr %0, null
  br i1 %1, label %cacheIsNull, label %cont

cacheIsNull:                                      ; preds = %entry
  %2 = call swiftcc %swift.metadata_response @"$s7SwiftUI13_TaskModifierVMa"(i64 255) #19
  %3 = extractvalue %swift.metadata_response %2, 0
  %4 = extractvalue %swift.metadata_response %2, 1
  %5 = call ptr @swift_getWitnessTable(ptr @"$s7SwiftUI13_TaskModifierVAA04ViewD0AAMc", ptr %3, ptr undef) #9
  store atomic ptr %5, ptr @"$s7SwiftUI13_TaskModifierVAcA04ViewD0AAWL" release, align 8
  br label %cont

cont:                                             ; preds = %cacheIsNull, %entry
  %6 = phi ptr [ %0, %entry ], [ %5, %cacheIsNull ]
  ret ptr %6
}
```

Reply:
This is some interesting type you shared. It doesn't even seem to be public. Could you share more about it?

Reply:
Let me try to collect use cases and I will ping you offline.

Reply:
So it seems that the PWT symbol is not synthesised directly when dealing with PATs and when protocol is defined in a different module then the conformance.

Reply:
Does this mean if PAT is declared in a framework, the bindings need to retrieve protocol witness table during runtime?

Reply:
Not only PAT. Any protocol. I think we will have to use the conformance descriptor after all.

Reply:
Library evolution mode: https://www.swift.org/blog/library-evolution/
> Finally, to cope with adding new protocol requirements, protocol witness tables require runtime instantiation. Instead of emitting a witness table in the client code directly, the compiler emits a symbolic description of the conformance. The instantiation process places the protocol requirements in the correct order and fills missing entries to point to their default implementation, to produce a well-formed witness table which can be passed off to a dispatch thunk.